src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/LspServices.cs

```diff
@@ -11,22 +11,55 @@
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer;
 
-internal class LspServices : ILspServices
+internal sealed class LspServices : ILspServices
 {
+    private sealed class EmptyServiceProvider : IServiceProvider
+    {
+        public object? GetService(Type serviceType) => null;
+    }
+
+    public static LspServices Empty { get; } = new(new EmptyServiceProvider());
+
     private readonly IServiceProvider _serviceProvider;
-    public bool IsDisposed = false;
+
+    private readonly object _disposeLock = new();
+    private bool _isDisposed = false;
+
+    public bool IsDisposed => _isDisposed;
 
     public LspServices(IServiceCollection serviceCollection)
     {
         serviceCollection.AddSingleton<ILspServices>(this);
+        serviceCollection.AddSingleton<LspServices>(this);
+
         _serviceProvider = serviceCollection.BuildServiceProvider();
-        // Create all startup services
+
+        // By requesting the startup services, we ensure that they are created.
+        // This gives them an opportunity to set up any necessary state or perform.
         _serviceProvider.GetServices<IRazorStartupService>();
     }
 
-    public ImmutableArray<Type> GetRegisteredServices()
+    private LspServices(IServiceProvider serviceProvider)
     {
-        throw new NotImplementedException();
+        _serviceProvider = serviceProvider;
+    }
+
+    public void Dispose()
+    {
+        lock (_disposeLock)
```

Comment:
❓I've not seen this pattern often. Is there a reason we need to use a lock here? Is this because `Interlocked.CompareExchange` doesn't have a bool overload?

Reply:
Yeah, this is just a simple bit of state that has zero contention. I just wanted a lock to protect the value, since the previous implementation was so strange. (i.e., the previous implementation won't even set `IsDisposed` to true unless `_serviceProvider is IDisposable` is true.)

FWIW, there are plenty of places where we use just plain ol' locks in both [Razor]((https://github.com/search?q=repo%3Adotnet%2Frazor+%22lock+%28%22&type=code)) and [Roslyn](https://github.com/search?q=repo%3Adotnet%2Froslyn%20%22lock%20(%22&type=code). Trying to be lock free here just seemed like overkill.